buildSrc/subprojects/testing/src/main/kotlin/org/gradle/testing/DistributionTest.kt

```diff
@@ -0,0 +1,145 @@
+        current.getName() + " " + System.getProperty("os.arch")
+    }
+
+    @Internal
+    val gradleInstallationForTest: GradleInstallationForTestEnvironmentProvider = GradleInstallationForTestEnvironmentProvider(project)
+
+    @Internal
+    val binaryDistributions: BinaryDistributions = BinaryDistributions(project.layout)
+
+    @Internal
+    val libsRepository: LibsRepositoryEnvironmentProvider = LibsRepositoryEnvironmentProvider(project.layout)
+
+    init {
+        jvmArgumentProviders.add(gradleInstallationForTest)
+        jvmArgumentProviders.add(BinaryDistributionsEnvironmentProvider(binaryDistributions))
+        jvmArgumentProviders.add(libsRepository)
+    }
+}
+
+fun <K, V> Map<K, V>.asSystemPropertyJvmArguments(): Iterable<String> = this.map { (key, value) -> "-D$key=$value" }
+
+class LibsRepositoryEnvironmentProvider(layout: ProjectLayout) : CommandLineArgumentProvider, Named {
+
+    @Internal
+    val dir: DirectoryProperty = layout.directoryProperty()
+
+    @Input
+    var required: Boolean = false
+
+    override fun asArguments(): Iterable<String> =
+        if (required) mapOf("integTest.libsRepo" to dir.asFile.get().absolutePath).asSystemPropertyJvmArguments() else emptyList()
+
+    override fun getName(): String = "libsRepository"
+}
+
+class GradleInstallationForTestEnvironmentProvider(project: Project) : CommandLineArgumentProvider, Named {
+
+    @Internal
+    val gradleHomeDir: DirectoryProperty = project.layout.directoryProperty()
+
+    @Internal
+    val gradleUserHomeDir: DirectoryProperty = project.layout.directoryProperty()
+
+    @Internal
+    val toolingApiShadedJarDir: DirectoryProperty = project.layout.directoryProperty()
+
+    /**
+     * The user home dir is not wiped out by clean.
+     * Move the daemon working space underneath the build dir so they don't pile up on CI.
+     */
+    @Internal
+    val daemonRegistry: DirectoryProperty = project.layout.directoryProperty()
+
+    @Nested
+    val gradleDistribution = GradleDistribution(project, gradleHomeDir)
+
+    override fun asArguments(): Iterable<String> =
+        mapOf(
+            "integTest.gradleHomeDir" to gradleHomeDir.asFile.get().absolutePath,
+            "integTest.gradleUserHomeDir" to gradleUserHomeDir.asFile.get().absolutePath,
+            "org.gradle.integtest.daemon.registry" to daemonRegistry.asFile.get().absolutePath,
+            "integTest.toolingApiShadedJarDir" to toolingApiShadedJarDir.getAsFile().get().absolutePath
+        ).asSystemPropertyJvmArguments()
+
+    override fun getName(): String = "gradleInstallationForTest"
+}
+
+class BinaryDistributions(layout: ProjectLayout) {
+
+    @Input
+    var binZipRequired: Boolean = false
+
+    @Input
+    var distributionsRequired: Boolean = false
+
+    @InputDirectory
+    @PathSensitive(PathSensitivity.RELATIVE)
+    val distsDir: DirectoryProperty = layout.directoryProperty()
+
+    @Internal
+    lateinit var distZipVersion: String
```

Comment:
Why `lateinint` over a property?

Reply:
You mean `Property<String>`?

Reply:
Something like that, yes. Just trying to understand why/when one is chosen over the other.

This same question comes up in plugin API design.